File: server/src/main/java/au/com/codeka/warworlds/server/handlers/HandlerServlet.java
package au.com.codeka.warworlds.server.handlers;

import java.util.List;
import java.util.regex.Matcher;

import javax.servlet.GenericServlet;
import javax.servlet.ServletRequest;
import javax.servlet.ServletResponse;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

import au.com.codeka.warworlds.common.Log;

/**
 * Base {@link javax.servlet.Servlet} for working with handlers.
 */
public class HandlerServlet extends GenericServlet {
  private Log log = new Log("HandlerServlet");

  private final List<Route> routes;

  protected HandlerServlet(List<Route> routes) {
    this.routes = routes;
  }

  @Override
  public void service(ServletRequest request, ServletResponse response) {
    for (Route route : routes) {
      Matcher matcher = route.matches((HttpServletRequest) request);
      if (matcher != null) {
        handle(matcher, route, (HttpServletRequest) request, (HttpServletResponse) response);
        return;
      }
    }

    log.info(
        String.format(
            "Could not find handler for URL: %s",
            ((HttpServletRequest) request).getPathInfo()));
    ((HttpServletResponse) response).setStatus(404);
  }

  protected void handle(
      Matcher matcher,
      Route route,
      HttpServletRequest request,
      HttpServletResponse response) {
    try {
      RequestHandler handler = route.createRequestHandler();
      handler.setup(matcher, route.getExtraOption(), request, response);
      handler.handle();
    } catch (RequestException e) {
      e.populate(response);
    }
  }
}

File: server/src/main/kotlin/au/com/codeka/warworlds/server/handlers/HandlerServlet.kt
package au.com.codeka.warworlds.server.handlers

import au.com.codeka.warworlds.common.Log
import java.util.regex.Matcher
import javax.servlet.GenericServlet
import javax.servlet.ServletRequest
import javax.servlet.ServletResponse
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

/**
 * Base [javax.servlet.Servlet] for working with handlers.
 */
open class HandlerServlet protected constructor(private val routes: List<Route>)
  : GenericServlet() {
  private val log = Log("HandlerServlet")
  override fun service(request: ServletRequest, response: ServletResponse) {
    for (route in routes) {
      val matcher = route.matches(request as HttpServletRequest)
      if (matcher != null) {
        handle(matcher, route, request, response as HttpServletResponse)
        return
      }
    }
    log.info(String.format(
        "Could not find handler for URL: %s",
        (request as HttpServletRequest).pathInfo))
    (response as HttpServletResponse).status = 404
  }

  protected open fun handle(
      matcher: Matcher,
      route: Route,
      request: HttpServletRequest,
      response: HttpServletResponse) {
    try {
      val handler = route.createRequestHandler()
      handler.setup(matcher, route.extraOption, request, response)
      handler.handle()
    } catch (e: RequestException) {
      e.populate(response)
    }
  }
}
